### PR TITLE
Map Matter range to Tapo range for hue, saturation and brightness

### PR DESCRIPTION
**Pull request by MonicaisHer**
This PR implements a mapping of values from the Matter controller range to the Tapo bulbs range, affecting hue, saturation, and brightness/level. 

Here are the ranges:

| Parameter | Matter range | Tapo range |
|-----------|--------------|------------|
| Hue       | 0-254        | 0-359      |
| Saturation| 0-254        | 0-100      |
| Brightness/Level| [3-254](https://github.com/project-chip/connectedhomeip/blob/master/docs/guides/chip_tool_guide.md#step-7-control-application-data-model-clusters)   | 1-100      |

The following references are also used to validate and determine the above ranges:

- https://github.com/project-chip/connectedhomeip/tree/master/src/app/tests/suites
- https://docs.espressif.com/projects/esp-matter/en/main/esp32/developing.html
- https://github.com/project-chip/connectedhomeip/blob/b8389773bcb81b444b5ad62a68046af6f5ef4148/src/app/tests/suites/certification/Test_TC_CC_4_5.yaml#L82

### Testing
Hue:
```
# set default saturation
$ sudo chip-tool colorcontrol move-to-saturation 254 0 0 0 110 1
[tapo] 192.168.178.61: set color 359, 100

$ sudo chip-tool colorcontrol move-to-hue 0 1 0 0 0 110 1
[tapo] 192.168.178.61: set color 0, 100                                     <------------------- 0=>0

$ sudo chip-tool colorcontrol move-to-hue 1 1 0 0 0 110 1
# bridge logs
[tapo] 192.168.178.61: set color 1, 100                                     <------------------- 1=>1

$ sudo chip-tool colorcontrol move-to-hue 200 1 0 0 0 110 1
[tapo] 192.168.178.61: set color 282, 100                                <------------------- 200=>282

$ sudo chip-tool colorcontrol move-to-hue 254 1 0 0 0 110 1
[tapo] 192.168.178.61: set color 359, 100                                <------------------- 254=>359
```
Saturation:
```
# first set default hue
$ sudo chip-tool colorcontrol move-to-hue 254 1 0 0 0 110 1
[tapo] 192.168.178.61: set color 359, 78

$ sudo chip-tool colorcontrol move-to-saturation 0 0 0 0 110 1
[tapo] 192.168.178.61: set color 359, 0                                     <------------------- 0=>0

$ sudo chip-tool colorcontrol move-to-saturation 1 0 0 0 110 1
# bridge logs
[tapo] 192.168.178.61: set color 0, 0                                          <------------------- 1=>0

$ sudo chip-tool colorcontrol move-to-saturation 200 0 0 0 110 1
[tapo] 192.168.178.61: set color 359, 78                                     <------------------- 200=>78

$ sudo chip-tool colorcontrol move-to-saturation 254 0 0 0 110 1
[tapo] 192.168.178.61: set color 359, 100                                   <------------------- 254=>100
```
Brightness:
```
$ sudo chip-tool levelcontrol move-to-level 3 0 0 0 110 1
[tapo] 192.168.178.61: set brightness level 1                        <------------------- 3=>1
$ sudo chip-tool levelcontrol move-to-level 200 0 0 0 110 1
[tapo] 192.168.178.61: set brightness level 78                    <------------------- 200=>78

$ sudo chip-tool levelcontrol move-to-level 254 0 0 0 110 1
[tapo] 192.168.178.61: set brightness level 100                 <------------------- 254=>100
```

### CODE REVIEW

File: lighting.py
Comment (farshidtz) [inline lighting.py:111]: If this is changes to uint, should the following code still treat `value` as an array?

File: lighting.py
Comment (farshidtz) [inline lighting.py:130]: To make it more efficient:
```suggestion
                set_level(math.trunc(value * (100/254)))
```
and in 2 more places...

Comment (MonicaisHer) [reply lighting.py:130]: 2 lines would improve readability and make it clearer that we are converting matter value to Tapo value. in terms of code efficiency, it shouldn't really matter because the compiler should combine the two lines anyways. what do you think?

Comment (farshidtz) [reply lighting.py]: I don't think the compiler does that optimization. There is some minimal performance cost which is okay but the readability can be improved by adding comments too.

File: lighting.py
Comment (farshidtz) [inline lighting.py:77]: What is this for? 
The input is an integer: https://python-reference.readthedocs.io/en/latest/docs/ints/

**Comment by MonicaisHer** [comment]
Although most cases pass the manual testing, when the converted Tapo value is 0, value[0] return an empty value, which cause the following setting failed. An example:
```
$ sudo chip-tool levelcontrol move-to-level 1 0 0 0 110 1
# bridge logs
[callback] cluster=8 attr=0 value=[1, 95, 236, 255, 91, 45, 127]
[callback] level=1
[tapo] 192.168.178.61: set brightness level 0
Exception ignored on calling ctypes callback function: <function attributeChangeCallback at 0x7f2d6b7e81f0>
Traceback (most recent call last):
  File "/home/mengyi/Desktop/matter-tapo-mapping/matter-bridge-tapo/lighting.py", line 127, in attributeChangeCallback
    set_level(math.trunc(value[0] * (100/254)))
  File "/home/mengyi/Desktop/matter-tapo-mapping/matter-bridge-tapo/lighting.py", line 67, in set_level
    dev.setBrightness(level)
  File "/home/mengyi/connectedhomeip/out/python_env/lib/python3.10/site-packages/PyP100/PyL530.py", line 41, in setBrightness
    errorMessage = self.errorCodes[str(errorCode)]
KeyError: '-1008'
[1682071585.422214][4556:4566] CHIP:DMG: Endpoint 1, Cluster 0x0000_0008 update version to 71ef3cdf
[callback] cluster=8 attr=1 value=[]
[callback] Error: unhandled cluster 8 or attribute 1
```

**Comment by farshidtz** [comment]
You are showing two errors.

The first one is passing brightness 0 to the library as shows in message: `192.168.178.61: set brightness level 0` but the library is refusing it. This could indicate that a valid brightness range doesn't include 0, need to verify this.

The second one shows a call for an unhandled cluster and attribute.

**Comment by MonicaisHer** [comment]
I tested the brightness level using the library and validated that its range is from 1 to 100. In the case of hue and saturation, I added a check for the value parameter to set them to 0 when left empty (represented as `value=[]`). This allowed me to ensure that their respective ranges of 0-359 and 0-100 are now functioning properly.

The second error appears unrelated to or block this PR, can be addressed separately with another PR.

File: lighting.py
Comment (farshidtz) [inline lighting.py:109]: Please add comments, otherwise it will be hard to understand why this was done later.

**Review by farshidtz — APPROVED** [review]
Looks good. Thanks for the fixes.

**Comment by MonicaisHer** [comment]
> > I tested the brightness level using the library and validated that its range is from 1 to 100. In the case of hue and saturation, I added a check for the value parameter to set them to 0 when left empty (represented as `value=[]`). This allowed me to ensure that their respective ranges of 0-359 and 0-100 are now functioning properly.
> 
> Just for my understanding, if the valid range is from 1 to 100, why do you still want to send 0-100 to the device? Isn't 0 going to trigger the error from the library?

I am not attempting to send a value of 0-100 to the brightness level as it only accepts values from 1-100 and will result in an error if I send 0. 

Rather, I am sending a value of 0-100 to the saturation level, soI have added a condition that checks whether a value exists or not, and if not, sets the value to bytes([0]).

**Comment by farshidtz** [comment]
> I am not attempting to send a value of 0-100 to the brightness level as it only accepts values from 1-100 and will result in an error if I send 0.
> 
> Rather, I am sending a value of 0-100 to the saturation level, soI have added a condition that checks whether a value exists or not, and if not, sets the value to bytes([0]).

The code you added in https://github.com/farshidtz/matter-bridge-tapo/pull/3/commits/2f2644b889ebc43464a201932386884fc7973c2d applies to both brightness and saturation and so `set_level(0)` gets called, passing 0 to the library!

File: lighting.py
Comment (farshidtz) [reply lighting.py:108]: It will be hard to understand such code because it doesn't belong here. Such conditional work needs to happen in the right scope, before calling the device function or inside them. Defer to https://github.com/farshidtz/matter-bridge-tapo/issues/4

**Pull request merged**
